src/utils/async.js

```diff
@@ -10,6 +9,69 @@ export function delay<T>(callback: () => T): Promise<T> {
 export const sleep = (ms: number = 0): Promise<void> =>
   new Promise(resolve => setTimeout(resolve, ms));
 
+type BackoffMachine = {
+  waitsCompleted: () => number,
+  wait: () => Promise<void>,
+};
+
+/**
+ * Makes a machine that can sleep for increasing durations. The machine should
+ * be created before a loop starts, and .wait() should be called in each
+ * iteration of the loop. The machine should not be re-used after exiting the
+ * loop.
+ *
+ * The .waitsCompleted() getter can be used if the caller needs to implement
+ * "give up" logic for a network request by breaking out of the loop after a
+ * certain number of tries.
+ *
+ * The popular exponential backoff strategy is to increase the duration
+ * exponentially with the number of sleeps completed, with a base of 2, until  a
+ * ceiling is reached. E.g., if firstDuration is 100 and durationCeiling is 10 *
+ * 1000 = 10000, the sequence is
+ *
+ * 100, 200, 400, 800, 1600, 3200, 6400, 10000, 10000, 10000, ...
+ *
+ * Instead of using this strategy directly, we also apply "jitter." We use
+ * capped exponential backoff for the *upper bound* on a random duration, where
+ * the lower bound is always zero. Mitigating "bursts" is the goal of any
+ * "jitter" strategy, and the larger the range of randomness, the smoother the
+ * bursts. Keeping the lower bound at zero maximizes the range while preserving
+ * a capped exponential shape on the expected value. Greg discusses this in more
+ * detail in #3841.
+ */
+export const makeBackoffMachine = (): BackoffMachine => {
+  const firstDuration = 100;
+  const durationCeiling = 10 * 1000;
+  const base = 2;
+
+  let startTime: number | void;
+  let waitsCompleted: number = 0;
+
+  return {
+    waitsCompleted() {
+      return waitsCompleted;
+    },
+
+    async wait(): Promise<void> {
+      if (startTime === undefined) {
+        startTime = Date.now();
+      }
+
+      const duration =
+        Math.random() // "Jitter"
+        * Math.min(
+          // Upper bound of random duration should not exceed durationCeiling
+          durationCeiling,
+          firstDuration * base ** waitsCompleted,
```

Comment:
We should avoid computing large powers of base when `waitsCompleted` is clearly large enough that you're gonna end up choosing `durationCeiling`.  More discussion here: https://chat.zulip.org/#narrow/stream/92-learning/topic/exponential.20backoff.20--.20with.20jitter/near/820056

Reply:
I believe this concern was resolved in https://chat.zulip.org/#narrow/stream/92-learning/topic/exponential.20backoff.20--.20with.20jitter/near/820655, but let me know if not. 🙂